exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/diagnosiskey/DiagnosisKeyDao.java

```diff
@@ -119,31 +122,59 @@ public boolean verify(int verificationId, String requestChecksum, long totalKeyC
         }
     }
 
-    public List<TemporaryExposureKey> fetchAvailableKeysForEfgs(long operationId) {
+    @Transactional
+    public Optional<FederationOutboundOperation> fetchAvailableKeysForEfgs(boolean retry) {
         LOG.info("Fetching queued keys not sent to efgs.");
-        String sql = "select key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, " +
-                "visited_countries, days_since_onset_of_symptoms, origin, consent_to_share " +
+        Timestamp timestamp = new Timestamp(Instant.now().toEpochMilli());
+        String sql = "with batch as ( " +
+                "select key_data " +
                 "from en.diagnosis_key " +
-                "where efgs_operation = :efgs_operation " +
-                "order by key_data";
+                "where sent_to_efgs is null and retry_count >= :min_retry_count and retry_count < :max_retry_count limit 5000 ) " +
+                "update en.diagnosis_key " +
+                "set sent_to_efgs = :timestamp, retry_count = retry_count + 1 " +
+                "where key_data in (select key_data from batch) " +
+                "returning key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, " +
+                "visited_countries, days_since_onset_of_symptoms, origin, consent_to_share";
+
+        List<TemporaryExposureKey> keys = new ArrayList<>(jdbcTemplate.query(sql, Map.of(
+                "min_retry_count", retry ? 1 : 0,
+                "max_retry_count", retry ? MAX_RETRY_COUNT : 1,
+                "timestamp", timestamp
+        ), (rs, i) -> mapKey(rs)));
+
+        if (keys.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(new FederationOutboundOperation(keys, operationDao.startOperation(OperationDao.EfgsOperationDirection.OUTBOUND, timestamp)));
+        }
+    }
 
-        return new ArrayList<>(jdbcTemplate.query(sql, Map.of("efgs_operation", operationId), (rs, i) -> mapKey(rs)));
+    @Transactional
+    public void setNotSent(FederationOutboundOperation operation) {
+        String sql = "update en.diagnosis_key set sent_to_efgs = null where key_data in (:key_datas)";
+        jdbcTemplate.update(sql, Map.of(
+                "key_datas", operation.keys.stream().map(key -> key.keyData).collect(Collectors.toList())
+        ));
+        operationDao.markErrorOperation(operation.operationId, Optional.of(operation.batchTag));
     }
 
     @Transactional
+    public void resolveOutboundCrash() {
+        List<Timestamp> crashed = operationDao.getAndResolveCrashed(OperationDao.EfgsOperationDirection.OUTBOUND);
+
+        if (!crashed.isEmpty()) {
+            String sql = "update en.diagnosis_key set sent_to_efgs = null, retry_count = 0 where sent_to_efgs in (:timestamp)";
+            jdbcTemplate.update(sql, Map.of(
```

Comment:
These extra linebreaks look funny to my eye when this wouldn't even be a long row as a single-liner.

Reply:
yep, it was a little bit. I changed that to oneliner.